Document integer data types and ranges in Lua API

Wuzzy2 opened the pull request:
This a large documentation PR about integers. It took me almost a whole day to write.

It updates the `lua_api.md` documentation significantly by adding a distinction between "number" and "integer" in the documentation, and documenting value ranges.

## Rationale

Many Luanti functions and data structures expect integers and the C++ code can only accept a limited value range. The Lua API documentation rarely mentions these restrictions which has led to some confusion in the past.

Worse, Lua programmers passing floating-point or out-of-range values to functions or data structures that expect integers can potentially lead to bugs without the programmer realizing, because it was never documented.

Real example: In one game, I have found a `stack_max` of 99999, which is not a supported value but the programmer can't be blamed because the upper bound was not specified there.

The goal of this PR is to make Lua code more robust longterm by mentioning our number restrictions upfront.

## Overview

This PR is structured in multiple parts:

1. Section "Numbers and integers" was added where terminology and writing conventions are introduced. The internal representation of the Lua `number` as C++ `double` is mentioned here (see `LUA_NUMBER` in `lib/lua/src/luaconf.h`)
2. Add the word "integer" everywhere in the documentation
3. Attempt to "standardize" the syntax of integer ranges as `[min, max]`
4. Add permissible ranges where there were missing
5. Add a "safe integer range" of `[-2^53, 2^53]` in the Numbers and Integers section for the few cases where no range was mentioned. This safe integer range is the range that Luanti will now officially guarantee that all integers in this range can be represented in Lua numbers without "holes". This information is relevant for builtin functions
6. A few minor typos were fixed if they happened to be close to my other edits

 
The information about the integer types and value ranges is derived from the C++ code.

## Noteworthy details for review

To the reviewers I want to bring special attention to these points:

1. I was rather paranoid with value ranges and sometimes picked the most "pessimistic" ranges (read: smallest allowed range)
    1. ~~I picked a "pessimistic" range of `[-2^37, 2^37]` for the safe integer range. This one is derived from the lowest guaranteed value of `DBL_MANT_DIG` in `<float.c>`.~~ (raised to `[-2^53, 2^53]`)
    2. ~~I mentioned that if IEEE-754 floating-point numbers are supported, the safe integer range is `[-2^53, 2^53]`. Why 53? Because 53 is the 52-bit mantissa of double-precision IEEE-754 floats plus 1 extra bit because the specification implies there is an extra `1` bit at the beginning of the mantissa that is not actually stored.~~ (no longer relevant)
    3. ~~(I don't know if Luanti ALWAYS has IEEE-754 floats (thus making this distinction moot) or if it is possible to compile Luanti with lower float ranges)~~ (the answer is, for all practical intents and purposes, yes)
    4. For `set_int` of metadata storage, I made the value range explicitly 32-bit, rather than this just being a suggestion
    5. For inventory list sizes, I picked a much lower range than the C++ code indicates because in my tests, Luanti becomes very angry and laggy if I use extremely large sizes. So this range is picked exclusesively as a precaution
2. ~~Oddly, a C++ comment for the implementation of `core.set_node_level` claims the allowed range is only [0, 63] although other parts of the documentation indicate that it can go up to 127 and in my tests the high values work as well. Should I fix the C++ comment?~~ (comment removed)
3. ~~`star_seed` in `player:set_stars` is internally a 64-bit unsigned integer, which I guess is larger than Lua 5.1 can represent safely in integers. I intentionally did not mention a range here so that the safe integer range kicks in~~ (`star_seed` range now specified as `[u32]` just to keep our lives simpler lol)
4. For the rest, just review the changes. :stuck_out_tongue_closed_eyes: 

### Things to test

Although I have proofread my changes myself by using the C++ code as a reference, it is possible that still errors have snuck in. Please compare the documentation updates against the C++ code as well.

## Status

This PR is ready for review.

## Off-topic: Things to keep in mind for later

1. ~~IMO the C++ type of the internal stored value after you call `meta:set_int` should be switched from `int` to `long` because of the *de-facto* usage of 32-bit integers and the fact the documentation already suggests 32-bit integers. Technically, `int` only guarantees 16-bit values in the C++ spec~~ (This point is moot, Luanti refuses to compile on systems wheere `int` is 16-bit.)
2. It may make sense to define `math.mininteger` and `math.maxinteger` to define the safe integer range, so that programmers know *exactly* which numbers are "safe" to use as integer and at which point integers are "skipped" due to floating-point arithmetic. The names `math.mininteger` and `math.maxinteger` are stolen from Lua 5.3. :wink:

appgurueu commented:
Thanks for documenting this! I will try to take a comprehensive look soon.

Some thoughts:

For all practical purposes, we can assume IEEE-754. But the more limiting factor is usually going to be a cast to some 32-bit (or even smaller) integer type on the C++ side. This is what I wrote on the matter in the other repo: https://docs.luanti.org/for-creators/lua-conventions/#integers

> [set_int:] Technically, this is wrong because it is internally stored as an int which guarantees only 16-bit.

Then let's make it use `s32`, no? (That said, `int` is basically always 32 bit in practice.)
Spec should certainly guarantee at least 32 bits.

wmikita commented on doc/lua_api.md:
```diff
@@ -54,6 +54,61 @@ the build directory. For system-wide builds on Linux the share path is usually a
+### Implementation details
+
+In Luanti, Lua numbers are equivalent to the data type `double`
+in the C programming language (which is the default Lua behavior).
```
2 Lua mentions?

**suggestion:**
> In Lua 5.1 (and subsequently in Luanti), numbers are represented by the C `double` data type.

Wuzzy2 commented:
All direct code comments by @wmikita addressed. Thank you for looking at the small details, it is important to be precise.

> BTW, you could mention that some Lua 5.1 APIs are ambiguous on whether they take any real number or an integer and that behavior may vary there too from implementation to implementation (and to look out for that), e.g. with when floats are passed to math.random in PUC Lua 5.1 and LuaJIT 2.1. maybe (and I'm totally not doing this to substitute cat picture payments to @appgurueu)

Not sure, this might be overkill for a reference. I don't think we have to also document every Lua quirk as well.

> you could mention that there are tools like [Runtime Strictness](https://content.luanti.org/packages/LMD/strictest/) to catch undefined behaviors in both Lua and Luanti API?

Not sure about that, I don't know how useful this tool is. Maybe a good recommentation in docs.luanti.org tho (if it's a good tool).

-----

Question to everyone important in Luanti:

Is it safe to assume that in Luanti, `double` is a IEEE-754 double-precision floating point number? Or in other words: Does Luanti care enough to support systems where `double` is NOT based on IEEE-754 (for some reason)?
If the answer (to the 2nd question) is "no", I will increase the "safe integer range" to `[-2^53, 2^53]` and simplify the documentation a little. Please note this PR would then add to the `lua_api.md` a *promise* that this larger range works (instead of the current smaller range of `[-2^37, 2^37]`).

j-r commented on doc/lua_api.md:
Perhaps add `integer` here somewhere, too?

Wuzzy2 commented:
* `PcgRandom` seed and seq range reduced to `[ulua]` (from `[u64]`)
* Clarified correct use of `seed` in mapgen settings
* Increased safe integer range to `[-2^53, 2^53]` (from `[-2^37, 2^37]`); documentation now assumes that Lua numbers are IEEE-754 double precision floating point numbers because systems where that's not the case don't seem to supported by Luanti anyway (discussed in IRC)

appgurueu commented on doc/lua_api.md:
```diff
@@ -841,16 +848,20 @@ The mask is applied using binary AND.
-with dimensions w, h (integer, in tiles).
+* `<w>`: sheet width in tiles (integer [imagesize])
+* `<h>`: sheet height in tiles (integer [imagesize])
+* `<x>`: x position in tiles (integer [u32], 0-indexed)
```
same as for verticalframe, i think these should be imagesize

appgurueu commented on doc/lua_api.md:
```diff
@@ -75,40 +77,52 @@ All integer values have a range with a defined minimum and maximum.
-    [uint16] = [0,     2^16-1] = [0, 65535]
-    [uint32] = [0,     2^32-1] = [0, 4294967295]
-    [uint64] = [0,     2^64-1] = [0, 18446744073709551615]
+    [s16]  = [-2^15, 2^15-1] = [-32768, 32767]
```
could include `s8` / `u8` here

Wuzzy2 replied:
Spelling out small numbers (lower than 1000) explicitly was intentional for readability. So no s8/u8.

appgurueu commented on doc/lua_api.md:
```diff
@@ -75,40 +77,52 @@ All integer values have a range with a defined minimum and maximum.
-that cannot store negative integers)
+(s = "signed", ranges that include negative integers;
+ u = "unsigned", ranges that don't include negative integers;
+ these terms are borrowed from the C language)
```
very much a nitpick: the C reference confuses me a bit. it's `uint<x>_t` and `int<x>_t` in `stdint.h`.

`s` and `u` are just common abbreviations that tend to get used, in our case they happen to be defined by `irrTypes.h`.

Wuzzy2 replied:
Read again: The TERMS "signed" and "unsigned" are borrowed from C, not the abbreviations u16, u32, etc.

appgurueu commented on doc/lua_api.md:
```diff
@@ -654,9 +668,9 @@ on top of `cobble.png`.
-* `<p>`: current animation frame (counting starts at 0)
+* `<t>`: tile count (in each direction) (integer [imagesize])
+* `<n>`: animation frame count (integer [imagesize])
+* `<p>`: current animation frame (integer [0, 2^31-1], counting starts at 0)
```
should probably be imagesize as well, anything larger is always out of bounds

Wuzzy2 replied:
Going by your logic it would actually be [0, 22999] instead, because counting starts at 0.

appgurueu commented on doc/lua_api.md:
```diff
@@ -6109,8 +6132,8 @@ Utilities
   image and return it in string form.
-    * `width`: Width of the image (integer, range [0, 4294967295])
-    * `height`: Height of the image (integer, range [0, 4294967295])
+    * `width`: Width of the image (integer in range [u32])
```
the code probably does not require this at the moment because it sidesteps any actual image classes, but just to be safe, i think we should also restrict this to imagesize.

because, well, this is an image, and it makes little sense to encode images with > 20k pixels in each direction this way...

Wuzzy2 replied:
Rejected.

This is just encoding, not rendering and in my own tests this function manages to deal with very large pixel values just fine (but it will of course take more time). We should expect Luanti to be able to manage large amounts of data as well unless there's a bug.

`[imagesize]` is speficially for texture modifiers only and was added because Luanti itself added this restriction due to technical constraints (not performance).

appgurueu commented on doc/lua_api.md:
```diff
@@ -6772,32 +6796,34 @@ Environment access
+        * `water_level` (integer [s16])
+        * `flags` (string, this is `mg_flags` in `core.get_mapgen_settings`)
+    * **WARNING**: `seed` is broken and should not be used. Internally, mapgen
+        seeds are actually in the [u64] in Luanti. So if the actual mapgen
```
> seeds are actually in the [u64] _range_ in Luanti. So if the actual mapgen


it looks like suggest change is broken on GitHub at the moment :(

appgurueu reviewed (commented):
looked at the changes since my last review, looks good, i have some minor comments

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -668,9 +672,9 @@ on top of `cobble.png`.
-* `<p>`: current animation frame (counting starts at 0)
+* `<t>`: tile count (in each direction) (integer [imagesize])
+* `<n>`: animation frame count (integer [imagesize])
+* `<p>`: current animation frame (integer [0, 2^31-1], counting starts at 0)
```
counting starts at 0 tho, but imagesize at 1

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -827,16 +844,20 @@ The mask is applied using binary AND.
 `<color>` is specified as a `ColorString`.
-`<ratio>` is an int ranging from 0 to 255 or the word "`alpha`". If
-it is an int, then it specifies how far to interpolate between the
+`<ratio>` is an integer in range [0, 255] or the string `"alpha"`. If
```
No, the quotation marks are generally accepted as string delimiters in this documentation.

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -75,40 +77,52 @@ All integer values have a range with a defined minimum and maximum.
+ u = "unsigned", ranges that don't include negative integers;
+ these terms are borrowed from the C language)
+
+The [slua] range is the *safe integer range* in Lua. This contains all
```
Oops. It turns out my claim was not just misleading, but mathematically wrong ("all integers"). Fixing …

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -88,6 +88,7 @@ and may be abbreviated like so:
     [u32]  = [0,     2^32-1] = [0, 4294967295]
```
You're indeed wrong. A bare `[x]` in Markdown is safe. What you mean are links, and those are written as `[link text](https://some_url.example/index.html)`.

Fun fact: I originally used `[int32]` and `[uint32]` before but decided to scrap that in favor of identifiers that reflect how the types look in Luanti C++ (IrrlichtMt).

Wuzzy2 commented:
OK, I've removed the comments in `l_env.cpp` and added a brief explainer at the top.
I didn't touch the other related files because otherwise this PR becomes too bloated (a lot of files would have to be touched and it's kinda out-of-scope already). I suggest to do the rest separately.

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -1800,17 +1892,17 @@ where `x` and `y` range from `-1` to `1`, with `0` being central. `-1` is
-precisely positioned items in the HUD.
+The `offset` field specifies a pixel offset from the position, specified
+as integers. Contrary to position, the offset is not scaled to screen size.
+This allows for some precisely positioned items in the HUD.
```
Oops, it turns out that offset actually uses floats.

sfan5 commented on doc/lua_api.md:
```diff
@@ -54,6 +54,78 @@ the build directory. For system-wide builds on Linux the share path is usually a
 Paths on other operating systems will differ.
 
+Numbers and integers
+--------------------
```
should we have have some language in this section to say that it is informational?
so effectively that a function taking a specific value range will not mean that it will also take this value range in future versions.

Wuzzy2 commented on doc/lua_api.md:
```diff
@@ -54,6 +54,78 @@ the build directory. For system-wide builds on Linux the share path is usually a
+
+### Implementation details
+
+In the Luanti Lua API, numbers are internally represented by the `double`
```
Reworded it to "In the Lua runtime used by Luanti …".

appgurueu commented on doc/lua_api.md:
```diff
@@ -8865,10 +9016,10 @@ child will follow movement and rotation of that bone.
+    * `count`: number of items, must be an integer in range [1, 32]
     * If `count` exceeds the `"main"` list size, the list size will be used instead.
-* `hud_get_hotbar_itemcount()`: returns number of visible items
+* `hud_get_hotbar_itemcount()`: returns amount of visible items
```
personally i found the previous form about equally clear, but i also don't really see any harm in the change, so i'm fine either way :shrug:

appgurueu commented:
> I also fixed a false claim stating this function is not a hash function because there are no collisions. But this is false.

The phrasing can be improved, but the claim has a point. The term "hash function" is very misleading. This does not do any of the things you normally expect from hash functions (jumble things so they're spread out randomly), and it does have properties which you normally don't expect (it is reversible).

As such, the better term to use would be "encoding" or "bit packing". I don't like calling this "perfect hashing". Every injective function - every encoding - could be called a perfect hash function by this line of reasoning.
That term would only really be useful if we had a 2^48 slot hash table which nobody has.

So I suggest phrasing along the lines of: "This has historically been referred to as a hash function,
but it is more accurate to call it an encoding since it is reversible (see `core.get_position_from_hash`)."

Wuzzy2 commented:
Rebased!

sfan5 commented:
Needs a rebase (sorry). We'll merge it after this.

Wuzzy2 commented:
Rebase completed!
